src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java

```diff
@@ -1389,6 +1464,87 @@ else if (path.toFile().getName().contains("sun")) {
         }
     }
 
+    private static String[] validateRuleNameOrder(String fileName, String lastRuleName,
+                                                  String[] lastRuleNameParts, String ruleName) {
+        final boolean[] beforeFirstSpace = {true};
+        final Long ruleNumberPartsAmountLong = ruleName.chars()
+                .peek(character -> {
+                    if (character == ' ') {
+                        beforeFirstSpace[0] = false;
+                    }
+                })
+                .filter(character -> beforeFirstSpace[0] && character == '.')
+                .count() + 1;
+        final int ruleNumberPartsAmount = ruleNumberPartsAmountLong.intValue();
+        final String[] ruleNameParts = ruleName.split("[. ]", ruleNumberPartsAmount + 1);
+
+        if (lastRuleName != null) {
+            final int lastRuleNumberPartsAmount = lastRuleNameParts.length - 1;
+            boolean lastRuleNumberPartWasEqual = false;
+            final String outOfOrderReason = fileName + " rule '" + ruleName
+                    + "' is out of order compared to '" + lastRuleName + "'";
+            int partIndex;
+            for (partIndex = 0; partIndex < ruleNumberPartsAmount; partIndex++) {
+                if (lastRuleNumberPartsAmount <= partIndex) {
+                    // equal up to here and last rule has less parts,
+                    // thus order is correct, stop comparing
+                    break;
+                }
+
+                final String ruleNumberPart = ruleNameParts[partIndex];
+                final String lastRuleNumberPart = lastRuleNameParts[partIndex];
+                final boolean ruleNumberPartsAreNumeric = IntStream.concat(
+                        ruleNumberPart.chars(),
+                        lastRuleNumberPart.chars()
+                ).allMatch(Character::isDigit);
+
+                if (ruleNumberPartsAreNumeric) {
+                    final int numericRuleNumberPart = parseInt(ruleNumberPart);
+                    final int numericLastRuleNumberPart = parseInt(lastRuleNumberPart);
+                    Assert.assertThat(
+                            outOfOrderReason,
+                            numericRuleNumberPart < numericLastRuleNumberPart,
+                            describedAs("'%0' should not be less than '%1'",
+                                    is(false),
+                                    numericRuleNumberPart, numericLastRuleNumberPart));
+                }
+                else {
+                    Assert.assertThat(
+                            outOfOrderReason,
+                            ruleNumberPart.compareToIgnoreCase(lastRuleNumberPart) > 0,
+                            describedAs("'%0' should not be greater than '%1'",
```

Comment:
We don't have non-numeric sections, so I say this should just fail instead of allowing it and telling the user this rule is missing a number.

Reply:
Well, we don't have it in the sun and google styles.
But I thought I make it future-proof right away.
You could in the future add some style rules that have numberings like `1.a`, `1.b`, `1.c`, and so on.
As it was trivially to right away support this, I added it, don't you think it makes sense?

Reply:
Ok, lets keep it.